### PR TITLE
[Text field] New component for Android

### PR DESCRIPTION
**Pull request by ayush547**
### Platforms Impacted
- [ ] iOS
- [ ] macOS
- [ ] win32 (Office)
- [ ] windows
- [x] android

### Description of changes

Created new component "Text Field" based on fluent mobile guidelines.

### Verification

| Unfocused | Focused |
|----------------------------------------------|--------------------------------------------|
| Screenshot_2023-03-29-19-51-40-02_3238b23fbbdef38ccf4a465a34706bab | Screenshot_2023-03-29-19-54-11-57_3238b23fbbdef38ccf4a465a34706bab |

This PR creates the basic "Text Field" implementation for Android.
Basic jest test added. E2E tests will be added in a separate PR along with documentation and focused mode different icon support.

Input runs successfully on iOS, but will not be added to tester until validations by iOS crew is carried out.

https://user-images.githubusercontent.com/49569838/230850423-a8ec89c2-c5e4-458e-b9c1-e1d1fb9b45db.mov

### Pull request checklist

This PR has considered (when applicable):
- [ ] Automated Tests
- [ ] Documentation and examples
- [ ] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

### CODE REVIEW

File: packages/components/Input/src/Input.tsx [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+              {accessoryIcon && (
+                <Slots.accessoryIconPressable
+                  onTouchStart={(e) => (accessoryButtonOnPress ? accessoryButtonOnPress(e) : input.props.setText(''))}
+                >
+                  <Slots.accessoryIcon {...accessoryIconProps} accessible={false} />
+                </Slots.accessoryIconPressable>
+              )}
+            </Slots.input>
+          </IconWrapper>
+          {assistiveText && <Slots.assistiveText>{error ? error : assistiveText}</Slots.assistiveText>}
```
Comment (rohanprasadofficial) [inline packages/components/Input/src/Input.tsx:103]: Check for RTL issues once if any.

**Comment by harrieshin** [comment]
why do we need secondary text at all? i haven't seen that in fluent mobile design

**Comment by lyzhan7** [comment]
What state do you expect iOS to be in for this PR? From the PR title it seems like this is a new component for iOS as well as Android, but the description mentions this is primarily for Android.

File: apps/fluent-tester/src/FluentTester.tsx [hunk 2]
```diff
@@ -208,6 +208,7 @@ export const FluentTester: React.FunctionComponent<FluentTesterProps> = (props:
   const TestComponentView: React.FunctionComponent = () => {
     return (
       <ScrollView
+        keyboardShouldPersistTaps="handled"
```
Comment (ankraj12) [inline apps/fluent-tester/src/FluentTester.tsx:211]: Could you add a small note for why this was required as part of Input changes?

File: apps/fluent-tester/src/TestComponents/Input/InputDefault.tsx [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import OutlineSvg from '../../../assets/outlineIcon.svg';
+
+export const dismissIconProps: IconProps = { svgSource: { src: DismissSvg, viewBox: '0 0 20 20' } };
+export const outlineIconProps: IconProps = { svgSource: { src: OutlineSvg, viewBox: '0 0 20 20' } };
+export const filledIconProps: IconProps = { svgSource: { src: FilledSvg, viewBox: '0 0 20 20' } };
+
+export const InputDefault: React.FunctionComponent = () => {
+  return (
+    <View>
+      <Input value="Value prop passed!" />
```
Comment (ankraj12) [inline apps/fluent-tester/src/TestComponents/Input/InputDefault.tsx:18]: Consider expanding the tester code to demonstrate usage of value component and error handling props - yo can choose it as a separate PR

File: packages/components/Input/package.json [hunk 4]
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "@fluentui-react-native/input",
+  "version": "0.1.0",
+  "description": "add component-description",
```
Comment (ankraj12) [inline packages/components/Input/package.json:4]: supernit

**Comment by ayush547** [comment]
> What state do you expect iOS to be in for this PR? From the PR title it seems like this is a new component for iOS as well as Android, but the description mentions this is primarily for Android.

Secondary text conversation is still open for iOS. Other than that everything _should_ work for iOS. Care was taken to include changes that work for iOS (like [cursorColor prop](https://reactnative.dev/docs/textinput#cursorcolor-android) (Android only) vs [selectionColor prop](https://reactnative.dev/docs/textinput#selectioncolor) (Both iOS and Android). 
That said, the component hasn't been tested for iOS (keyboard / focus behavior specifically is a bit notorious)!

File: apps/fluent-tester/src/TestComponents/Input/InputDefault.tsx (hunk 3 above)
Comment (ayush547) [reply apps/fluent-tester/src/TestComponents/Input/InputDefault.tsx]: Updated tests

File: packages/components/Input/package.json (hunk 4 above)
Comment (ayush547) [reply packages/components/Input/package.json:4]: Updated

File: packages/components/Input/package.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    "lint": "fluentui-scripts eslint",
+    "test": "fluentui-scripts jest",
+    "update-snapshots": "fluentui-scripts jest -u",
+    "prettier": "fluentui-scripts prettier",
+    "prettier-fix": "fluentui-scripts prettier --fix true"
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/microsoft/fluentui-react-native.git",
+    "directory": "packages/components/ComponentTemplate"
```
[the remark this answers is not in the document]
Comment (ayush547) [reply packages/components/Input/package.json]: Missed this, have replaced with correct path

**Comment by harrieshin** [comment]
> > What state do you expect iOS to be in for this PR? From the PR title it seems like this is a new component for iOS as well as Android, but the description mentions this is primarily for Android.
> 
> Secondary text conversation is still open for iOS. Other than that everything _should_ work for iOS. Care was taken to include changes that work for iOS (like [cursorColor prop](https://reactnative.dev/docs/textinput#cursorcolor-android) (Android only) vs [selectionColor prop](https://reactnative.dev/docs/textinput#selectioncolor) (Both iOS and Android). That said, the component hasn't been tested for iOS (keyboard / focus behavior specifically is a bit notorious)! 

I don't think this PR title or description should mention iOS at this point of time, because it isn't properly tested and confirm it is ready for v1 usage on ios. lets then scope it for Android? at the minimum we should just make sure all the CI works to launching this demo page on ios.

File: apps/fluent-tester/src/TestComponents/Input/InputTest.tsx [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    testID: INPUT_TESTPAGE,
+    component: InputDefault,
+  },
+];
+
+export const InputTest: React.FunctionComponent = () => {
+  const status: PlatformStatus = {
+    win32Status: 'Backlog',
+    uwpStatus: 'Backlog',
+    iosStatus: 'Experimental',
```
Comment (ankraj12) [inline apps/fluent-tester/src/TestComponents/Input/InputTest.tsx:20]: Dropping a point for review here - if we go ahead with iOS validations for basic structure, this holds true - else we would like to update this back to Backlog.

File: packages/components/Input/src/Input.types.ts [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+export interface InputProps extends PressableFocusProps {
+  /*
+   * Source URL or name of the icon to show on the input.
+   */
+  icon?: IconSourcesType;
+
+  /**
+   * The width of the button.
```
Comment (ankraj12) [inline packages/components/Input/src/Input.types.ts:136]: what is "button" referenced here for?

File: packages/components/Input/src/Input.types.ts [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+   * Source URL or name of the accessory icon to show on the input.
+   * Shown while input has text in it. Not shown while input is used as a control component.
+   *
+   * @default - dismiss icon is shown.
+   * Set as null to disabled this icon.
+   */
+  accessoryIcon?: IconSourcesType | null;
+
+  /**
+   * A callback to call on accessoryIcon click event.
```
Comment (ankraj12) [inline packages/components/Input/src/Input.types.ts:150]: ,maybe "Callback that is called when the text input's accessoryIcon is clicked" - refencing the RN handler docs

File: packages/components/Input/src/Input.types.ts [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+  /**
+   * Determines which keyboard to open.
+   * enum("default", 'numeric', 'email-address', "ascii-capable", 'numbers-and-punctuation', 'url', 'number-pad', 'phone-pad', 'name-phone-pad',
+   * 'decimal-pad', 'twitter', 'web-search', 'visible-password')
+   * Determines which keyboard to open, e.g.numeric.
+   * The following values work across platforms: - default - numeric - email-address - phone-pad
+   * The following values work on iOS: - ascii-capable - numbers-and-punctuation - url - number-pad - name-phone-pad - decimal-pad - twitter - web-search
+   * The following values work on Android: - visible-password
+   */
+  type?: KeyboardTypeOptions | undefined;
```
Comment (ankraj12) [inline packages/components/Input/src/Input.types.ts:206]: Can you help understand why you did not re-use [KeyBoardType](Callback that is called when the text input's submit button is pressed.

) for this

File: packages/components/Input/src/Input.types.ts [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+
+  /**
+   * Placeholder to display in the input.
+   */
+  placeholder?: string;
+
+  /**
+   * These props facilitate behaviour changes of the text field.
+   */
+  textInputProps?: Omit<TextInputProps, 'defaultValue' | 'value' | 'keyboardType' | 'onChangeText' | 'onBlur' | 'onFocus' | 'placeholder'>;
```
Comment (ankraj12) [inline packages/components/Input/src/Input.types.ts:177]: based on discussion, are you considering to change this to a base class?

File: packages/components/Input/src/InputTokens.ts [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    spacingIconContent: 16,
+    paddingHorizontal: 16,
+    spacingInputVertical: 12,
+    spacingAssistiveTextVertical: 4,
+    spacingLabelTop: 12,
+    spacingInputSecondary: 8,
+    spacingAccessoryIconStart: 8,
+    spacingAssistiveTextStart: 0,
+    spacingLabelStart: 0,
+    accessoryIconSize: 24,
```
Comment (ankraj12) [inline packages/components/Input/src/InputTokens.ts:24]: Do we not have any tokens defined for this values?

File: packages/components/Input/src/__tests__/Input.test.tsx [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+import * as renderer from 'react-test-renderer';
+
+import { Input } from '../Input';
+
+describe('Input component tests', () => {
+  it('Input default', () => {
+    const tree = renderer.create(<Input accessoryIcon={null} />).toJSON();
+    expect(tree).toMatchSnapshot();
+  });
+
```
Comment (ankraj12) [inline packages/components/Input/src/__tests__/Input.test.tsx:13]: Please add a note to track any more tests that you might want to add on this.

**Review by ankraj12 — COMMENTED** [review]
Can you please share what if any (what) validations you have for this - I could see addition in E2E test, iOS and a11y testing.

**Review by ankraj12 — CHANGES_REQUESTED** [review]
Can you please confirm on any pending validations on this apart from E2E test add, iOS validations and a11y testing.

File: apps/fluent-tester/src/FluentTester.tsx (hunk 2 above)
Comment (ayush547) [reply apps/fluent-tester/src/FluentTester.tsx:211]: Added

File: apps/fluent-tester/src/TestComponents/Input/InputDefault.tsx [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+        accessoryIcon={dismissIconProps}
+        label="Label"
+        assistiveText="Assistive Text"
+        secondaryText="Secondary"
+      />
+      <Input
+        error={error}
+        icon={outlineIconProps}
+        placeholder="Only text up to 5 characters!"
+        accessoryIcon={dismissIconProps}
```
[the remark this answers is not in the document]
Comment (ayush547) [reply apps/fluent-tester/src/TestComponents/Input/InputDefault.tsx:36]: Updated

File: packages/components/Input/src/Input.types.ts (hunk 8 above)
Comment (ayush547) [reply packages/components/Input/src/Input.types.ts:150]: Have followed similar language to that used already in our repo (like onClick for button)

File: packages/components/Input/src/Input.types.ts (hunk 7 above)
Comment (ayush547) [reply packages/components/Input/src/Input.types.ts:136]: Replaced, thanks for carching

File: packages/components/Input/src/Input.types.ts (hunk 9 above)
Comment (ayush547) [reply packages/components/Input/src/Input.types.ts:206]: Keyboardtype is used under the hood, exposed as "type" to match fluent web props, added same as comment

File: packages/components/Input/src/InputTokens.ts (hunk 11 above)
Comment (ayush547) [reply packages/components/Input/src/InputTokens.ts:24]: We do not have tokens for icon size, updated spacing tokens

File: apps/fluent-tester/src/TestComponents/Input/InputTest.tsx (hunk 6 above)
Comment (ayush547) [reply apps/fluent-tester/src/TestComponents/Input/InputTest.tsx]: Changing iOS to backlog based on offline discussions.

File: packages/components/Input/src/Input.types.ts (hunk 10 above)
Comment (ayush547) [reply packages/components/Input/src/Input.types.ts:177]: Picking this up later

**Comment by ankraj12** [comment]
> > > What state do you expect iOS to be in for this PR? From the PR title it seems like this is a new component for iOS as well as Android, but the description mentions this is primarily for Android.
> > 
> > 
> > Secondary text conversation is still open for iOS. Other than that everything _should_ work for iOS. Care was taken to include changes that work for iOS (like [cursorColor prop](https://reactnative.dev/docs/textinput#cursorcolor-android) (Android only) vs [selectionColor prop](https://reactnative.dev/docs/textinput#selectioncolor) (Both iOS and Android). That said, the component hasn't been tested for iOS (keyboard / focus behavior specifically is a bit notorious)!
> 
> I don't think this PR title or description should mention iOS at this point of time, because it isn't properly tested and confirm it is ready for v1 usage on ios. lets then scope it for Android? at the minimum we should just make sure all the CI works to launching this demo page on ios.



> > > What state do you expect iOS to be in for this PR? From the PR title it seems like this is a new component for iOS as well as Android, but the description mentions this is primarily for Android.
> > 
> > 
> > Secondary text conversation is still open for iOS. Other than that everything _should_ work for iOS. Care was taken to include changes that work for iOS (like [cursorColor prop](https://reactnative.dev/docs/textinput#cursorcolor-android) (Android only) vs [selectionColor prop](https://reactnative.dev/docs/textinput#selectioncolor) (Both iOS and Android). That said, the component hasn't been tested for iOS (keyboard / focus behavior specifically is a bit notorious)!
> 
> I don't think this PR title or description should mention iOS at this point of time, because it isn't properly tested and confirm it is ready for v1 usage on ios. lets then scope it for Android? at the minimum we should just make sure all the CI works to launching this demo page on ios.

Agreed that the component can not be called V1 ready on iOS unless the crew works on it. I think we can keep the component enabled in iOS given it loads the page, looks right and is functional. It would help build for mobile more broadly in initial state by developers - instead of android only/ iOS only to start. Let us know if that sounds right/ if you have a differing opinion.  @ejlayne - for any inputs.

File: packages/components/Input/src/Input.types.ts [hunk 14]
```diff
@@ -203,7 +203,7 @@ export interface InputProps extends PressableFocusProps {
    * The following values work on iOS: - ascii-capable - numbers-and-punctuation - url - number-pad - name-phone-pad - decimal-pad - twitter - web-search
    * The following values work on Android: - visible-password
    */
-  type?: KeyboardTypeOptions | undefined;
+  type?: KeyboardTypeOptions | undefined; // Uses KeyBoardType under the hood, exposed as type to match fluent web
```
Comment (ankraj12) [inline packages/components/Input/src/Input.types.ts:206]: nit: remove this

**Review by ankraj12 — APPROVED** [review]

**Pull request merged**